src/utils/indexer.hpp

```diff
@@ -120,35 +134,94 @@ struct Indexer {
     std::tuple<Ts...> idxs;
     (
         [&] {
-          std::get<Is>(idxs) = idx / std::get<Is>(N);
-          idx -= std::get<Is>(idxs) * std::get<Is>(N);
-          std::get<Is>(idxs) += std::get<Is>(start);
+          std::get<Is>(idxs) = idx / N[Is];
+          idx -= std::get<Is>(idxs) * N[Is];
+          std::get<Is>(idxs) += start[Is];
         }(),
         ...);
     return idxs;
   }
 
   template <std::size_t... Is>
-  KOKKOS_FORCEINLINE_FUNCTION static std::array<int, sizeof...(Ts)>
-  GetFactors(std::tuple<Ts...> Nt, std::index_sequence<Is...>) {
-    std::array<int, sizeof...(Ts)> N;
+  KOKKOS_FORCEINLINE_FUNCTION void
+  GetIndicesCArrayImpl(int idx, int *indices, std::index_sequence<Is...>) const {
+    (
+        [&] {
+          indices[Is] = idx / N[Is];
+          idx -= indices[Is] * N[Is];
+          indices[Is] += start[Is];
+        }(),
+        ...);
+  }
+
+  template <std::size_t... Is>
+  KOKKOS_FORCEINLINE_FUNCTION Kokkos::Array<int, sizeof...(Ts)>
+  GetIndicesKArrayImpl(int idx, std::index_sequence<Is...>) const {
+    Kokkos::Array<int, sizeof...(Ts)> indices;
+    (
+        [&] {
+          indices[Is] = idx / N[Is];
+          idx -= indices[Is] * N[Is];
+          indices[Is] += start[Is];
+        }(),
+        ...);
+    return indices;
+  }
+
+  template <std::size_t... Is>
+  KOKKOS_FORCEINLINE_FUNCTION std::array<int, sizeof...(Ts)>
+  GetIndicesArrayImpl(int idx, std::index_sequence<Is...>) const {
+    std::array<int, sizeof...(Ts)> indices;
+    (
+        [&] {
+          indices[Is] = idx / N[Is];
+          idx -= indices[Is] * N[Is];
+          indices[Is] += start[Is];
+        }(),
+        ...);
+    return indices;
+  }
+
+  template <std::size_t... Is>
+  KOKKOS_FORCEINLINE_FUNCTION static Kokkos::Array<int, sizeof...(Ts)>
+  GetFactors(Kokkos::Array<int, sizeof...(Ts)> Nt, std::index_sequence<Is...>) {
+    Kokkos::Array<int, sizeof...(Ts)> N;
     int cur = 1;
     (
         [&] {
           constexpr std::size_t idx = sizeof...(Ts) - (Is + 1);
-          std::get<idx>(N) = cur;
-          cur *= std::get<idx>(Nt);
+          N[idx] = cur;
+          cur *= Nt[idx];
         }(),
         ...);
     return N;
   }
 
-  std::array<int, sizeof...(Ts)> N;
-  std::array<int, sizeof...(Ts)> start;
-  std::array<int, sizeof...(Ts)> end;
+  Kokkos::Array<int, sizeof...(Ts)> N;
+  Kokkos::Array<int, sizeof...(Ts)> start;
+  Kokkos::Array<int, sizeof...(Ts)> end;
+  std::size_t _size;
+};
```

Comment:
I switched these to `Kokkos::Array` to avoid an endless stream of `__host__` function called from `__device__ __host__` function warnings from `std::array`'s allocator and `operator[]`. I'm not familiar with the nuance between the two outside of being able to capture from device code or not.

Reply:
I am not aware of any functional difference between them (aside from capturability), so this looks good to me. In #1156, I had gone to filling C-style arrays as a way to deal with filling indices in the hierarchical loops into a single array.